07_web_endpoints/discord_bot.py

```diff
@@ -0,0 +1,230 @@
+# [here](https://discord.com/developers/docs/interactions/application-commands).
+
+# ## Defining the asgi app with modal
+
+# We now create a `POST /get_weather` endpoint using [FastAPI](https://fastapi.tiangolo.com/) and Modal's
+# [@asgi_app](https://modal.com/docs/guide/webhooks#serving-asgi-and-wsgi-apps) decorator to handle
+# interactions with our Discord app (so that every time a user does a slash
+# command, we can respond to it).
+
+# Let's get the imports out of the way and define an [`App`](https://modal.com/docs/reference/modal.App)
+
+import json
+
+import modal
+
+app = modal.App("example-discord-weather-bot")
+
+# We define an [Image](https://modal.com/docs/guide/images) that has the [`python-weather`](https://github.com/null8626/python-weather) package and
+# the [FastAPI](https://fastapi.tiangolo.com/) package installed.
+
+image = modal.Image.debian_slim(python_version="3.11").pip_install(
+    "python-weather==2.0.7", "fastapi[standard]==0.115.4", "pynacl==1.5.0"
+)
+
+# We define a function that uses the `python_weather` library to get the weather of a city
+# Note that since Discord requires an interaction response within 3 seconds, we
+# use [`spawn`](https://modal.com/docs/reference/modal.Function#spawn) to kick off
+# `get_weather_for_city`as a background task from the asgi app while returning a `defer` message to
+# Discord within the time limit.
+
+
+@app.function(image=image)
+async def get_weather_forecast_for_city(
+    city: str, interaction_token, application_id
+):
+    import aiohttp
+    from python_weather import IMPERIAL, Client, Error, RequestError
+
+    interaction_url = f"https://discord.com/api/v10/webhooks/{application_id}/{interaction_token}/messages/@original"
+    async with Client(unit=IMPERIAL) as client:
+        try:
+            weather = await client.get(city)
+            daily_forecasts = "\n".join(
+                [
+                    f"Date: {daily.date}, Highest temperature: {daily.highest_temperature}°F, Lowest Temperature: {daily.lowest_temperature}°F"
+                    for daily in weather
+                ]
+            )
+            message = f"The forecast for {weather.location} is as follows:\n{daily_forecasts}"
+        except RequestError:
+            message = "An error occurred, issue with connecting to weather api"
+        except Error:
+            message = "An error occurred, please check city name"
+
+    json_payload = {"content": message}
+    async with aiohttp.ClientSession() as session:
+        async with session.patch(interaction_url, json=json_payload) as resp:
+            print(await resp.text())
+
+
+# We now define an asgi app using the Modal ASGI syntax [@asgi_app](https://modal.com/docs/guide/webhooks#serving-asgi-and-wsgi-apps).
+
+
+@app.function(
+    secrets=[modal.Secret.from_name("weather-discord-secret")],
+    keep_warm=1,  # eliminates risk of container startup making discord ack time too long
```

Comment:
We should also `allow_concurrent_inputs` here -- no need to spin up a replica for every call from Discord.